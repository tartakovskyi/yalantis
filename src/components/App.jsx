import React, { useEffect, useState }  from 'react'
import { apiGetUsers } from "../api"
import Employees from './Employees'
import EmployeesBirthday from './EmployeesBirthday'


export const AppContext = React.createContext()


const App = () => {

  const [users, setUsers] = useState({})

  useEffect(() => {
    apiGetUsers()
    .then(response => {
      let usersByAlphabet = {A:[],B:[],C:[],D:[],E:[],F:[],G:[],H:[],I:[],J:[],K:[],L:[],M:[],N:[],O:[],P:[],Q:[],R:[],S:[],T:[],U:[],V:[],W:[],X:[],Y:[],Z:[]}
      response.data.forEach(user => {
        const firstLetter = user.lastName.substr(0,1)
        usersByAlphabet[firstLetter].push(user)
      })
      setUsers(usersByAlphabet)
    })
    .catch(error => {
      console.log(error)
    })
  }, [])

  return (
    <div className="App">
      <header>
        <div className="container">
          <a href="/">Yalantis React School</a>
        </div>  
      </header>
      <main>
        <div className="container">    
          <div className="row">
            {Object.keys(users).length 
              ?
              <AppContext.Provider value={{users}}>
                <Employees />
                <EmployeesBirthday />    
              </AppContext.Provider>
              :
              null   
            }
          </div>
        </div>
      </main>
    </div>
  )
}

export default App
