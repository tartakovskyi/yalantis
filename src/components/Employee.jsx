import React, { useContext, useState }  from 'react'
import { AppContext } from './App'


const Employee = ({ user:{ id, firstName, lastName }, letter}) => {

  const [active, setActive] = useState(0)
  const {changeUserStatus} = useContext(AppContext)

  const cls = active === 0 ? 'employee__name' : 'employee__name active'

  const onChangeValue = e => {
    const value = Number(e.target.value)
    setActive(value)
    changeUserStatus(id, letter, value)
  }

  return (
    <li className="employee">
      <h4 className={cls}>{firstName} {lastName}</h4>
      <form>
        <label className="form-check">
          <input 
            type="radio"
            name={`active_${id}`}
            value="0"
            id={`notActive_${id}`}
            className="form-check-input"
            onChange={onChangeValue}
            checked={active === 0}
          />
          <span className="form-check-label">not active</span>
        </label>
        <label className="form-check">  
          <input 
            type="radio"
            name={`active_${id}`}
            value="1"
            id={`active_${id}`}
            className="form-check-input"
            onChange={onChangeValue}
            checked={active === 1}
          />
          <span className="form-check-label">active</span>
        </label>
      </form>
    </li>
  )
}


export default Employee

