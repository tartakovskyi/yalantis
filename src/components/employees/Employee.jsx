import React, { useContext, useState } from "react";
import PropTypes from "prop-types";
import { AppContext } from "../App";

const Employee = ({ user: { id, firstName, lastName } }) => {
  const { activeUsers, changeUserStatus } = useContext(AppContext);
  const [active, setActive] = useState(activeUsers.includes(id) ? 1 : 0);

  const cls = active === 0 ? "employee__name" : "employee__name active";

  const onChangeValue = (e) => {
    const value = Number(e.target.value);
    setActive(value);
    changeUserStatus(id, value);
  };

  return (
    <li className="employee">
      <h4 className={cls}>
        {firstName} {lastName}
      </h4>
      <form>
        <label className="form-check">
          <input
            type="radio"
            name={`active_${id}`}
            value="0"
            id={`notActive_${id}`}
            className="form-check-input"
            onChange={onChangeValue}
            checked={active === 0}
          />
          <span className="form-check-label">not active</span>
        </label>
        <label className="form-check">
          <input
            type="radio"
            name={`active_${id}`}
            value="1"
            id={`active_${id}`}
            className="form-check-input"
            onChange={onChangeValue}
            checked={active === 1}
          />
          <span className="form-check-label">active</span>
        </label>
      </form>
    </li>
  );
};

Employee.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string.isRequired,
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
  }).isRequired,
};

export default Employee;
