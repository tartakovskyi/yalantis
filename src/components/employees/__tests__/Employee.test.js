import React from "react";
import { render } from "@testing-library/react";
import { AppContext } from "../../App";
import Employee from "../Employee";

test("Employee shold render correctly", () => {
	const user = {
		"id": "5e00928d91e7feaa9872ec08",
		"firstName": "Yang",
		"lastName": "Carson",
		"dob": "2019-02-26T16:52:36.244Z"
	}
  const activeUsers = ["5e00928dac694570760105c4","5e00928deee1b67fa4c44a2a","5e00928d1a9ac4bd431ff876","5e00928d4267f8de85992159","5e00928d8d97a843dbb05270","5e00928d3939c957e9c63c2a"]

  const changeUserStatus = () => {}
	
  const {container} = render(
    <AppContext.Provider value={{activeUsers, changeUserStatus}}>
      <Employee user={user} />
    </AppContext.Provider>
  )

	expect(container.firstChild).toMatchSnapshot();
});
