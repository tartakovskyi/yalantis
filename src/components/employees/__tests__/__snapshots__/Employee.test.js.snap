// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee shold render correctly 1`] = `
<li
  class="employee"
>
  <h4
    class="employee__name"
  >
    Yang
     
    Carson
  </h4>
  <form>
    <label
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="notActive_5e00928d91e7feaa9872ec08"
        name="active_5e00928d91e7feaa9872ec08"
        type="radio"
        value="0"
      />
      <span
        class="form-check-label"
      >
        not active
      </span>
    </label>
    <label
      class="form-check"
    >
      <input
        class="form-check-input"
        id="active_5e00928d91e7feaa9872ec08"
        name="active_5e00928d91e7feaa9872ec08"
        type="radio"
        value="1"
      />
      <span
        class="form-check-label"
      >
        active
      </span>
    </label>
  </form>
</li>
`;
